### ISSUE TITLE
Nullable reference types: Analysis of null checks on fields

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#9978]
I have been playing with nullable reference types, which is already working remarkably great! To try it out, I have [patched up](https://github.com/bkoelman/ResharperCodeContractNullability/tree/staticnull) my analyzer to convert Resharper's `(Item)NotNull/CanBeNullAttribute` annotations to nullable reference types. Then ran that on an [actual annotated codebase](https://github.com/bkoelman/DogAgilityCompetitionManagement). 

During that process I discovered some oddities, which I'll create issues for here.

**Version Used**: Built from source, branch NullableReferenceTypes (commit https://github.com/dotnet/roslyn/commit/8553cdac75c2b0901e8ae401aaa23022f0689935)

**Steps to Reproduce**:
The next block of code raises warning 8202: Possible dereference of a null reference:

``` csharp
public class Test
{
    private string? value;

    public void M()
    {
        if (value != null)
        {
            int i = value.Length; // CS8202
        }
    }
}
```

This can be worked around using:

``` csharp
    public void M()
    {
        string? valueSnapshot = value;
        if (valueSnapshot != null)
        {
            int i = valueSnapshot.Length; // no warning
        }
    }
```

**Expected Behavior**: The behavior is technically correct. Another thread could change the field value in-between. I cannot think of a better way, just wanted to mention this scenario, which gets somewhat annoying, as it occurs very often (in non-multithreaded code).

**Actual Behavior**:
CS8202 warning on the line with `value.Length` inside the if statement.

@AlekseyTs

**Comment by AlekseyTs** [comment]
At this point the behavior is expected, state of fields of reference types is not tracked.

**Comment by AlekseyTs** [comment]
CC @MadsTorgersen

**Comment by cston** [comment]
Fixed in #23063.
